### ISSUE TITLE
docs.rs for dependent crates seems broken

### ISSUE THREAD

**Issue by mbillingr** [serde-rs/serde#1295]
I just published a new version of a crate that depends on serde, and noticed that the documentation went missing. Here is the build log: https://docs.rs/crate/arff/0.2.0/builds/101512.

> Compiling serde v1.0.63
     Running `rustc --crate-name serde .cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.63/src/lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 --cfg 'feature="std"' --cfg 'feature="default"' -C metadata=f2d60ad5195a964a -C extra-filename=-f2d60ad5195a964a --out-dir /home/cratesfyi/cratesfyi/debug/deps -L dependency=/home/cratesfyi/cratesfyi/debug/deps --cap-lints allow --cfg de_boxed_c_str --cfg de_rc_dst --cfg integer128`
error[E0658]: 128-bit type is unstable (see issue #35118)

(excerpt)

I do not use 128 bit integers and did not enable them (at least not intentionally). Running the test suite and building the docs works on CI and locally. 

No idea what's going on there :) I'm not even really sure this is an issue with serde, but no other dependencies are affected as far I can tell.

Cheers

P.S.: Looks like other recently upgraded crates are affected too (e.g. [validator](https://docs.rs/crate/validator/0.7.0/builds/101514), [rkv](https://docs.rs/crate/rkv/0.3.0/builds/101516))

**Comment by hcpl** [comment]
Related commit: https://github.com/serde-rs/serde/commit/c81bab18ad76b4f753d06998619894271ad14c73.

I think the fix should have used a Cargo feature instead of a `rustc` cfg to allow downstream crates to use it when building on docs.rs. Like:

```toml
[package.metadata.docs.rs]
features = ["serde/docs_rs_temporary_workaround"]
```

**Comment by hcpl** [comment]
> I do not use 128 bit integers and did not enable them (at least not intentionally). Running the test suite and building the docs works on CI and locally. No idea what's going on there :)

@mbillingr if you don't mind, I'd prefer explaining this whole situation with a timeline (using YYYY.MM.DD format to eliminate confusions):

- 2018.03.16: the `rustc` version docs.rs is using now is `nightly-2018-03-16` (you can see that from build logs).
- 2018.03.27: `i128` support [has been stabilized](https://github.com/rust-lang/rust/pull/49101); this occured when nightly was 1.26.
- 2018.03.29: Rust 1.25 was released and Rust 1.26 became beta.
- 2018.05.10: Rust 1.26 was released which contains features people were _yearning_ for: 128-bit integers as primitive types, `impl Trait`, inclusive range `..=` syntax and more. For the full list of new features see https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1250-2018-03-29.
- Since 2018.05.10 till today: as the aforementioned features finally became stable, people started to use them in many places -- because they are very convenient. Which means bumping the minimal stable Rust version that is able to compile their code to 1.26.
  On the other side of the coin, libraries that work with numerical primitives _have_ to support new `{i,u}128` types. If not, these libraries will likely annoy those who expect the support already and confuse those not familiar with the history of Rust. This also produces a chain effect by blocking dependent libraries from accessing the new types at all and the more "core" the library is the more widespread the effect will be. `serde` is definitely one of these -- which prompted [implementing `i128` & `u128` support](https://github.com/serde-rs/serde/pull/1263) for it too.

Seems like the natural progression of a language to make code work better and look prettier -- and this makes people happy.

Alas, there is a roadblock: the `rustc` version on docs.rs hasn't been updated since 2018.03.16! The code uploaded to docs.rs justifiably assumes docs.rs to have at least the latest stable Rust, or even better, the latest nightly (it includes everything the latest stable is capable of). As you witnessed yourself, this is not the case. Many people have already reported this to https://github.com/onur/docs.rs/issues/23.

Thankfully, the work needed to fix this is ongoing: there is a PR filed to docs.rs repo to update its Rust version: https://github.com/onur/docs.rs/pull/207. But it it unknown when the PR gets merged. Time will tell.

**Comment by mbillingr** [comment]
@hcpl thank you for the detailed explanation. I had not realized docs.rs was using an old version of rustc (somehow focussing on the 1.26 part and missing the 2018-03-16).

I guess the issue can be closed, or maybe you want to leave it open for now as a landing pad for other misguided souls?

**Comment by dtolnay** [comment]
I published Serde 1.0.64 with a workaround that should make this work. To confirm, I published a version of serde_json and it seems to have built successfully in docs.rs. Thanks all!

Before: https://docs.rs/serde_json/1.0.18/serde_json/
After: https://docs.rs/serde_json/1.0.19/serde_json/

**Comment by hcpl** [comment]
@mbillingr anyone may copy around my comment (https://github.com/serde-rs/serde/issues/1295#issuecomment-392968504) to spread the information if they wish so. Or they can just link to the comment which also works.

@dtolnay oh wow -- https://github.com/serde-rs/serde/commit/34a7108b73475f3f4a5900019bd122a2c9c63085#diff-beade9a1819c15eaafbb1eb58aca7e3bR57 -- that's a nice trick, I like it.
